src/StreamJsonRpc/JsonMessageFormatter.cs

```diff
@@ -429,6 +513,23 @@ private JsonRpcError ReadError(JToken json)
             };
         }
 
+        private void ClearProgressObject(object requestId)
+        {
+            lock (this.requestProgressMap)
+            {
+                lock (this.progressMap)
+                {
```

Comment:
Be *very* careful when taking nested locks like this. You must be sure that *all* uses of either of these locks always lock in the same order. In this case, that means that you *never* acquire the progressMap lock followed by the requestProgressMap lock. Keep in mind that these locks might be taken individually in distinct methods making it harder to recognize. If they are taken in different orders you can easily deadlock.
It's almost always advisable to avoid two lock objects in a class to avoid the issue altogether. I recommend in this case you just pick one. If it doesn't make sense to pick one of the two dictionaries arbitrarily, it's quite common to define a `private readonly object syncObject = new object();` field and use that to lock.